library/std/src/panicking.rs

```diff
@@ -180,6 +164,69 @@ pub fn take_hook() -> Box<dyn Fn(&PanicInfo<'_>) + 'static + Sync + Send> {
     }
 }
 
+/// Atomic combination of [`take_hook`] + [`set_hook`].
+///
+/// [`take_hook`]: ./fn.take_hook.html
+/// [`set_hook`]: ./fn.set_hook.html
+///
+/// # Panics
+///
+/// Panics if called from a panicking thread.
+///
+/// Panics if the provided closure calls any of the functions [`panic::take_hook`],
+/// [`panic::set_hook`], or [`panic::update_hook`].
+///
+/// Note: if the provided closure panics, the panic will not be able to be handled, resulting in a
+/// double panic that aborts the process with a generic error message.
+///
+/// [`panic::take_hook`]: ./fn.take_hook.html
+/// [`panic::set_hook`]: ./fn.set_hook.html
+/// [`panic::update_hook`]: ./fn.update_hook.html
+///
+/// # Examples
+///
+/// The following will print the custom message, and then the normal output of panic.
+///
+/// ```should_panic
+/// #![feature(panic_update_hook)]
+/// use std::panic;
+///
+/// panic::update_hook(|prev| {
+///     Box::new(move |panic_info| {
+///         println!("Print custom message and execute panic handler as usual");
+///         prev(panic_info);
+///     })
+/// });
+///
+/// panic!("Custom and then normal");
+/// ```
+#[unstable(feature = "panic_update_hook", issue = "92649")]
+pub fn update_hook<F>(hook_fn: F)
+where
+    F: FnOnce(
+        Box<dyn Fn(&PanicInfo<'_>) + 'static + Sync + Send>,
+    ) -> Box<dyn Fn(&PanicInfo<'_>) + 'static + Sync + Send>,
+{
+    if thread::panicking() {
+        panic!("cannot modify the panic hook from a panicking thread");
+    }
+
+    unsafe {
+        let guard = HOOK_LOCK.write();
+        let old_hook = HOOK;
+        HOOK = Hook::Default;
+
+        let hook_for_fn = match old_hook {
+            Hook::Default => Box::new(default_hook),
+            Hook::Custom(ptr) => Box::from_raw(ptr),
+        };
+
+        let hook = hook_fn(hook_for_fn);
```

Comment:
How do you feel about the idea of changing this so that the output type of the `hook_fn` isn't a box and we box it ourselves after the closure returns? That way the usages of `take_hook` never have to call `Box::new`.

I guess doing that kinda ruins the usage with `set_hook`... maybe not the best idea vs maintaining consistency.

Reply:
Yeah, I tried to keep the same API as `set_hook`. Looking at the usages of `set_hook` in the compiler, I only see two kinds of calls:

* `set_hook(prev)`, where `prev` is the return of a previous call to `take_hook`
* `set_hook(Box::new(|info| { ... }))`, with custom code inside the closure

So I guess we can try to remove the explicit `Box` from `update_hook`.

As I mentioned in another comment it would be interesting to see if something like this is possible:

```rust
panic::update_hook(|prev, info| {
    println!("panic handler A");
    prev(info);
});
```

Which is nice because it removes one level of nested closures, and it would also handle the `box`ing internally.

Reply:
> As I mentioned in another comment it would be interesting to see if something like this is possible:

Do you mean something like this?

```rust
#[unstable(feature = "panic_update_hook", issue = "92649")]
pub fn update_hook<F>(hook_fn: F)
where
    F: Fn(
        &(dyn Fn(&PanicInfo<'_>) + 'static + Sync + Send), &PanicInfo<'_>,
    ) + Send + Sync + 'static,
{
    if thread::panicking() {
        panic!("cannot modify the panic hook from a panicking thread");
    }

    unsafe {
        let guard = HOOK_LOCK.write();
        let old_hook = HOOK;
        HOOK = Hook::Default;

        let hook_for_fn = match old_hook {
            Hook::Default => Box::new(default_hook),
            Hook::Custom(ptr) => Box::from_raw(ptr),
        };

        HOOK = Hook::custom(move |info| hook_fn(&hook_for_fn, info));
        drop(guard);
    }
}

impl Hook {
    fn custom(f: impl Fn(&PanicInfo<'_>) + 'static + Sync + Send) -> Self {
        Self::Custom(Box::into_raw(Box::new(f)))
    }
}
```

I ran into some lifetime issues too when trying this that I'm assuming are the same ones you ran into:

```
error[E0308]: mismatched types
   --> library/std/src/panicking.rs:228:43
    |
228 |         HOOK = Hook::Custom(Box::into_raw(hook));
    |                                           ^^^^ one type is more general than the other
    |
    = note: expected type `for<'r, 's> Fn<(&'r PanicInfo<'s>,)>`
               found type `Fn<(&PanicInfo<'_>,)>`
```

Which ended up being an issue with how `for<...>` is uninferrable, but adding a new constructor lets you hint to the compiler the correct bounds. I don't entirely understand how introducing the constructor is avoiding inference for the closure so I don't think I can sufficiently explain how this fixes things. I had to have a friend from the compiler team help me figure it out, lol...

Edit: I asked a bunch of questions and am going to attempt to re-explain this. Turns out closures aren't subject to the same inference rules as the rest of the language because "higher-rank inference is undecidable, or so the nerds tell me". If the hook closure is assigned to a variable it goes through inference and fails to infer the `for` bound correctly, but if its used directly as the argument in a function the compiler never needs inference, because it already knows the expected type when figuring out the closures type and so it's able to get it right.

Reply:
Was just working on this, I think it works if I use a helper function to create the closure:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e45977aea46d1147d307dbdd73ecc682

Reply:
hehe, yea, you nerd sniped me into wanting to see what was up with that lifetime error, quite a fun / nasty problem!

Reply:
So thanks for the snippet, that solution is more clear than what I had. I just updated the code and it seems to work. I left the original implementation in a separate commit, and I will add a short summary of the benefits and drawbacks of each approach to the tracking issue. So that if the community finds some use case that is not possible with this new approach, we can quickly update it.